Instructor: edit session: multiple choice questions: support generating options for all other students #8412

Haozhe321 opened the pull request:
Fixes #8412 

**Outline of Solution**
1. Added a new option `students (excluding self)`
2. Passes an instance of student into MCQ and MSQ so that this student's name can be excluded from the generated list of choices
3. Created new enum type in FeedbackParticipantType 
4. Change the method signature for some methods to take in an instance of student as mentioned in point 2. Since `FeedbackMsqQuestionDetails` and `FeedbackMcqQuestionDetails` inherits from `FeedbackQuestionDetails`, a change in the method signature of these 2 methods also resulted a change in the method signature of it's abstract method which must be overridden. Hence this method signature in some other subclasses classes also have to be overridden, but does not affect the other classes. 

<img width="1152" alt="8412" src="https://user-images.githubusercontent.com/26651556/36195362-2c1c94ca-11a8-11e8-9708-a11e92addbd3.png">

Haozhe321 commented:
![Uploading Screen Shot 2018-02-13 at 4.07.12 PM.png…]()

wkurniawan07 commented:
@Haozhe321 is this ready for review?

Haozhe321 commented:
@wkurniawan07 Yup it is!

wkurniawan07 commented on src/main/java/teammates/common/util/Templates.java:
```diff
@@ -233,9 +233,12 @@ public static String populateTemplate(String template, String... keyValuePairs)
             public static final String INSTRUCTOR_SELECTED = "${instructorSelected}";
             public static final String STUDENTS_TO_STRING = "${FeedbackParticipantType.STUDENTS.toString()}";
+            public static final String STUDENTS_EXCLUDING_SELF_TO_STRING = "${FeedbackParticipantType."
+                    + "STUDENTS_EXCLUDING_SELF.toString()}";
```
Rather than the awkward breaking at `.`, let's just move the entire String to the next line?

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/FeedbackParticipantType.java:
```diff
@@ -10,6 +10,8 @@
     SELF(true, true, false, "Feedback session creator (i.e., me)", "Giver (Self feedback)", ""),
     STUDENTS(true, true, true, "Students in this course", "Other students in the course", "Other students in the course"),
+    STUDENTS_EXCLUDING_SELF(false, true, true, "Students in this course", "Other students in the course",
+            "Other students in the course"),
```
The text displayed here is identical to the `STUDENTS` option, and that will be confusing to users.

Haozhe321 replied:
I'm not too sure what else to write for the text; it seems suitable for it's purpose. On another note, this new addition is not used in the feedback path so it won't be shown to the users.

wkurniawan07 replied:
@Haozhe321 you'd have noticed by now that it really is not the case. ([example](#diff-b3cdf5a28c730589262ac1b1a1db2720R1747))

Haozhe321 replied:
The link brings me back to the pull request. Is it supposed to be an image or a code snippet?

Haozhe321 replied:
![image](https://user-images.githubusercontent.com/26651556/36383613-a06c27dc-15c7-11e8-8d0a-87eccd140071.png)
Is it this one? 
Should I change to "Other students in the course (excluding self)"? But other students kind of already means excluding self.

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackConstantSumQuestionDetails.java:
```diff
@@ -224,7 +225,8 @@ public String getQuestionWithExistingResponseSubmissionFormHtml(
     public String getQuestionWithoutExistingResponseSubmissionFormHtml(
-            boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients) {
+            boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients,
+                StudentAttributes student) {
```
Applies to many other places: don't need the extra indentation here.

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -242,11 +248,24 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
+            optionList.sort(null);
+            break;
+
+        case STUDENTS_EXCLUDING_SELF:
```
This block looks almost identical to the `STUDENTS` one. Why not use the same block and only do extra logical steps when the option is `STUDENTS_EXCLUDING_SELF`?

Haozhe321 replied:
```
case STUDENTS:
    //fallthrough
case STUDENTS_EXCLUDING_SELF:
    List<StudentAttributes> studentList = StudentsLogic.inst().getStudentsForCourse(courseId);

    if(generateOptionsFor == FeedbackParticipantType.STUDENTS_EXCLUDING_SELF) {
        studentList.removeIf(studentInList -> studentInList.email.equals(studentDoingQuestion.email));
    }

    for (StudentAttributes student : studentList) {
        optionList.add(student.name + " (" + student.team + ")");
    }

    optionList.sort(null);
    break;
```

Is this okay?

wkurniawan07 commented on src/main/java/teammates/ui/controller/InstructorFeedbackPreviewAsStudentAction.java:
```diff
@@ -30,7 +30,7 @@ protected ActionResult execute() throws EntityDoesNotExistException {
         }
 
-        FeedbackSubmissionEditPageData data = new FeedbackSubmissionEditPageData(account, student, sessionToken);
+        FeedbackSubmissionEditPageData data = new FeedbackSubmissionEditPageData(account, previewStudent, sessionToken);
```
Not sure if this is something needs fixing; what is the effect of this?

Haozhe321 replied:
The `student` passed in here is null, so it results in a NPE when previewing the form as a student. Passing in previewStudent fixes this.

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -120,9 +121,14 @@ private void setMsqQuestionDetails(FeedbackParticipantType generateOptionsFor, S
         this.generateOptionsFor = generateOptionsFor;
-        this.numOfMsqChoices = generateOptionList(courseId).size();
-        Assumption.assertTrue("Can only generate students, teams or instructors",
+        if (generateOptionsFor.equals(FeedbackParticipantType.STUDENTS_EXCLUDING_SELF)) {
```
If I understand the change set correctly, you have added the implementation for `generateOptionList` for `STUDENTS_EXCLUDING_SELF`, which means this extra conditional should not be needed?

Haozhe321 replied:
In the original implementation, `generateOptionList(courseId).size()` would give an error because at this point of execution, the student is not passed into MsqQuestionDetails yet (the logic of execution for MCQ and MSQ seem to differ), so the method `generateOptionList` would result in NPE.

wkurniawan07 replied:
Leaving a note here that this should be related to #8415.

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -343,6 +366,20 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
+            break;
+
+        case STUDENTS_EXCLUDING_SELF:
+            List<StudentAttributes> studentListExcludingSelf = StudentsLogic.inst().getStudentsForCourse(courseId);
```
Similarly, this may be able to be compressed with the `STUDENTS` one.

Haozhe321 commented:
@wkurniawan07 
Changes made as requested. 
1) I've added UI tests for both mcq and msq
2) I've changed the booleans for STUDENTS_EXCLUDING_SELF to all `false` since this enum will not be used in the feedback path, it was created for the generation of options for mcq/msq.

Haozhe321 commented:
UI test changes made for the 2nd point mentioned above

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackRankRecipientsQuestionDetails.java:
```diff
@@ -106,7 +107,8 @@ public String getQuestionWithExistingResponseSubmissionFormHtml(
     public String getQuestionWithoutExistingResponseSubmissionFormHtml(
-            boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients) {
+            boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients,
+                StudentAttributes student) {
```
No need to indent this one further

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackRankQuestionDetails.java:
```diff
@@ -48,11 +49,12 @@ public boolean extractQuestionDetails(Map<String, String[]> requestParameters,
     public abstract String getQuestionWithoutExistingResponseSubmissionFormHtml(
-            boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients);
+            boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients,
+                StudentAttributes student);
```
No need to indent this one further

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackNumericalScaleQuestionDetails.java:
```diff
@@ -105,7 +106,8 @@ public String getQuestionWithExistingResponseSubmissionFormHtml(
     public String getQuestionWithoutExistingResponseSubmissionFormHtml(
-            boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients) {
+            boolean sessionIsOpen, int qnIdx, int responseIdx, String courseId, int totalNumRecipients,
+                StudentAttributes student) {
```
No need to indent this one further

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -328,6 +337,20 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
+    /**
+     * Get the number of MSQ choices when ticking the option "students (excluding self)".
+     * Which is equals to the number of students - 1
+     * @param courseId the courseId
```
No need for this if there is nothing to be explained.

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -328,6 +337,20 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
+     */
+    private int generateNumOfChoicesForStudentsExcludingSelf(String courseId) {
+
+        List<StudentAttributes> copyOfStudentList =
```
This is not really a `copyOf`, is it?

Haozhe321 replied:
I meant it as a a copy of the student list(in the liberal meaning of copy i.e. a copy of this paper). Is there a specific meaning to copyOf? 
Anyway changed it to `studentList`.

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMsqQuestionDetails.java:
```diff
@@ -328,6 +337,20 @@ public String getQuestionWithoutExistingResponseSubmissionFormHtml(
     }
 
+    /**
+     * Get the number of MSQ choices when ticking the option "students (excluding self)".
```
Don't mention `when ticking the option "students (excluding self)"`; how the method is being used is independent of what the method does.

Haozhe321 replied:
Just to be sure, so when writing Javadoc comments we only write what the method does?

wkurniawan07 replied:
Yes. You can optimize a method for its caller, but ultimately a method is its own construct and independent of its callers.

Haozhe321 commented:
@wkurniawan07 Changes made, and added new tests that you requested for. I added questions into the corresponding json file in order to generate the questions for clicking on the submission page, that's why there's a huge increase in the lines added because of the HTML files. 

Side node, only `FeedbackRubricQuestionUiTest` and `FeedbackRankQuestionUiTest` seem to be complete, the other questions are missing UI tests and their corresponding json data.

wkurniawan07 commented:
> Side node, only `FeedbackRubricQuestionUiTest` and `FeedbackRankQuestionUiTest` seem to be complete, the other questions are missing UI tests and their corresponding json data.

Your observation is not wrong, but it's the other way around: it's those two tests that are unnecessarily bulky (#3554). For this reason, if you're testing the submitting actions, can you instead do it in `StudentFeedbackSubmitPageUiTest`?

Haozhe321 commented:
@wkurniawan07 Changes made!

wkurniawan07 commented on src/test/java/teammates/test/cases/browsertests/StudentFeedbackSubmitPageUiTest.java:
```diff
@@ -594,6 +594,21 @@ private void testInputValidation() {
+        submitPage = loginToStudentFeedbackSubmitPage("Alice", "Open Session");
+        assertFalse(submitPage.doesMcqOptionExist(24, 0, "Alice Betsy</option></td></div>'\" (Team >'\"< 1</td></div>'\")"));
+        assertTrue(submitPage.doesMcqOptionExist(24, 0, "Charlie Davis (Team 2)"));
+        submitPage.submitWithoutConfirmationEmail();
```
Do you need to submit twice for this test case? i.e. can do the assertions for both question 24/25 (or 25/26 I'm not sure if it is 0- or 1-indexed) together before submitting?

Haozhe321 replied:
Yup can! Some part of the code on top submitted for every question and some submitted after a few questions so I wasn't sure which one to follow. Changed so that only one submission is required.

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java:
```diff
@@ -124,6 +124,24 @@ public void chooseMcqOption(int qnNumber, int responseNumber, String choiceName)
 
+    public boolean doesMcqOptionExist(int qnNumber, int responseNumber, String choiceName) {
+        try {
+            chooseMcqOption(qnNumber, responseNumber, choiceName);
```
Verifying a choice exists by choosing it is a kind of hack. Is there a cleaner method?

Haozhe321 replied:
How's the method below? 
The clicking action is no longer present. 
```
public boolean checkIfMcqOrMsqChoiceExists(int qnNumber, int responseNumber, String choiceName) {
    String name = Const.ParamsNames.FEEDBACK_RESPONSE_TEXT + "-" + qnNumber + "-" + responseNumber;
    name = SanitizationHelper.sanitizeStringForXPath(name);
    String sanitizedChoiceName = SanitizationHelper.sanitizeStringForXPath(choiceName);
    try{
        browser.driver.findElement(
                By.xpath("//input[@name=" + name + " and @value=" + sanitizedChoiceName + "]"));
    } catch(NoSuchElementException e) {
        return false;
    }
    return true;
}
```

Haozhe321 commented:
@wkurniawan07 changes made!

damithc commented:
@wenmogu Forgot to check: any changes to user docs needed?

wenmogu commented:
@damithc No change to user doc needed as there is not much elaboration about generating options from the student list.

wenmogu commented:
Sorry for the hasty comment I posted and deleted. This comment is based on my own observation of @Haozhe321 's branch. 

@damithc There is not much elaboration about generating options from the student list so to me the current explanation doesn't need a change in this case. The screenshot is also up-to-date on @Haozhe321 's branch. Maybe if @Haozhe321 wants to promote this new selection then can replace one of the screenshots with one showing the new option of `students (excluding self)` :)
